pkg/engine/query.go

```diff
@@ -0,0 +1,168 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package engine
+
+import (
+	"context"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/resource/plugin"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/fsutil"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"github.com/pulumi/pulumi/pkg/util/result"
+)
+
+type QueryOptions struct {
+	Events     eventEmitter // the channel to write events from the engine to.
+	Diag       diag.Sink    // the sink to use for diag'ing.
+	StatusDiag diag.Sink    // the sink to use for diag'ing status messages.
+	host       plugin.Host  // the plugin host to use for this query.
+	pwd, main  string
+	plugctx    *plugin.Context
+}
+
+func Query(ctx *Context, u UpdateInfo, opts UpdateOptions) result.Result {
+	contract.Require(u != nil, "update")
+	contract.Require(ctx != nil, "ctx")
+
+	defer func() { ctx.Events <- cancelEvent() }()
+
+	tracingSpan := func(opName string, parentSpan opentracing.SpanContext) opentracing.Span {
+		// Create a root span for the operation
+		opts := []opentracing.StartSpanOption{}
+		if opName != "" {
+			opts = append(opts, opentracing.Tag{Key: "operation", Value: opName})
+		}
+		if parentSpan != nil {
+			opts = append(opts, opentracing.ChildOf(parentSpan))
+		}
+		return opentracing.StartSpan("pulumi-query", opts...)
+	}("query", ctx.ParentSpan)
+	defer tracingSpan.Finish()
+
+	emitter, err := makeEventEmitter(ctx.Events, u)
+	if err != nil {
+		return result.FromError(err)
+	}
+
+	// First, load the package metadata and the deployment target in preparation for executing the package's program
+	// and creating resources.  This includes fetching its pwd and main overrides.
+	diag := newEventSink(emitter, false)
+	statusDiag := newEventSink(emitter, true)
+
+	proj, target := u.GetProject(), u.GetTarget()
+	contract.Assert(proj != nil)
+	contract.Assert(target != nil)
+
+	pwd, main, plugctx, err := ProjectInfoContext(&Projinfo{Proj: proj, Root: u.GetRoot()}, opts.host,
+		target, diag, statusDiag, tracingSpan)
+	if err != nil {
+		return result.FromError(err)
+	}
+
+	return query(ctx, u, QueryOptions{
+		Events:     emitter,
+		Diag:       diag,
+		StatusDiag: statusDiag,
+		host:       opts.host,
+		pwd:        pwd,
+		main:       main,
+		plugctx:    plugctx,
+	})
+}
+
+func newQuerySource(cancel context.Context, client deploy.BackendClient, u UpdateInfo,
+	opts QueryOptions) (deploy.QuerySource, error) {
+
+	allPlugins, _, err := installPlugins(client, u.GetProject(), opts.pwd,
+		opts.main,
+		u.GetTarget(), opts.plugctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Once we've installed all of the plugins we need, make sure that all analyzers and language plugins are
```

Comment:
I get that this is a copy-pasta'd comment, but we would never load provider plugins in query mode today, right?  (I am guessing we will in the future when we start to query live state...)